components/server/src/workspace/env-var-service.ts

```diff
@@ -0,0 +1,92 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { ProjectDB, UserDB } from "@gitpod/gitpod-db/lib";
+import {
+    CommitContext,
+    EnvVar,
+    EnvVarWithValue,
+    ProjectEnvVar,
+    User,
+    UserEnvVar,
+    WithEnvvarsContext,
+    Workspace,
+} from "@gitpod/gitpod-protocol";
+import { inject, injectable } from "inversify";
+import { ProjectsService } from "../projects/projects-service";
+
+export interface ResolvedEnvVars {
+    user: UserEnvVar[];
+    // all project env vars, censored included always
+    project: ProjectEnvVar[];
+    context: EnvVarWithValue[];
+    // merged workspace env vars
+    workspace: EnvVar[];
+}
+
+@injectable()
+export class EnvVarService {
+    @inject(UserDB)
+    private userDB: UserDB;
+
+    @inject(ProjectsService)
+    private projectsService: ProjectsService;
+
+    @inject(ProjectDB)
+    private projectDB: ProjectDB;
+
+    async resolve({ user, workspace }: { user: User | undefined; workspace: Workspace }): Promise<ResolvedEnvVars> {
+        const workspaceEnvVars = new Map<String, EnvVar>();
+        const merge = (envs: EnvVar[]) => {
+            for (const env of envs) {
+                workspaceEnvVars.set(env.name, env);
+            }
+        };
+
+        // 1. first merge user envs
+        let userEnvVars = user ? await this.userDB.getEnvVars(user.id) : [];
```

Comment:
Should we have prebuild guard here, instead of relying on passing `user` as null? i.e. a user can be passed always but will be ignored if workspace type is prebuild

Reply:
Since we are looking up the project from the workspace, we could also look up the user from it. And then also considering the workspace.type. So just passing in the workspace.

Reply:
Just to check you mean to use `workspace.ownerId`